src/Analyzers/CSharp/Analysis/RemoveUnnecessaryElseAnalyzer.cs

```diff
@@ -55,12 +57,39 @@ public static bool IsFixable(ElseClauseSyntax elseClause)
         if (!ifStatement.IsTopmostIf())
             return false;
 
-        StatementSyntax statement = ifStatement.Statement;
+        StatementSyntax ifStatementStatement = ifStatement.Statement;
 
-        if (statement is BlockSyntax block)
-            statement = block.Statements.LastOrDefault();
+        if (ifStatementStatement is not BlockSyntax ifBlock)
+            return CSharpFacts.IsJumpStatement(ifStatementStatement.Kind());
+        
+        if (elseClause.Statement is BlockSyntax elseBlock && LocalDeclaredVariablesOverlap(elseBlock, ifBlock, semanticModel))
+            return false;
+
+        var lastStatementInIf = ifBlock.Statements.LastOrDefault();
+
+        return lastStatementInIf is not null
+               && CSharpFacts.IsJumpStatement(lastStatementInIf.Kind());
+    }
+
+    private static bool LocalDeclaredVariablesOverlap(BlockSyntax elseBlock, BlockSyntax ifBlock, SemanticModel semanticModel)
+    {        
+        var elseVariablesDeclared = semanticModel.AnalyzeDataFlow(elseBlock)!
+            .VariablesDeclared;
+        
+        if (elseVariablesDeclared.IsEmpty)
+            return false;
+
+        var ifVariablesDeclared = semanticModel.AnalyzeDataFlow(ifBlock)!
+            .VariablesDeclared;
+        
+        if (ifVariablesDeclared.IsEmpty)
+            return false;
 
-        return statement is not null
-            && CSharpFacts.IsJumpStatement(statement.Kind());
+        var elseVariableNames = elseVariablesDeclared
+            .Select(s => s.Name)
+            .ToImmutableHashSet();
+        
+        return ifVariablesDeclared
+            .Any(s => elseVariableNames.Contains(s.Name));
```

Comment:
Same as comment above: better would be to use foreach loop to avoid allocations.